highfidelity/hifi
# [Case 6491] Remainder of entitySelectionTool clean up (details below).

This pull request contains the remainder, Parts 2 - 6, of the entitySelectionTool cleanup for [FogBugz Case 6491](https://highfidelity.fogbugz.com/f/cases/6491/edit-js-work-proposed-by-1st-playable).  Part 1 was merged via PR #11165.  This pull request also replaces https://github.com/highfidelity/hifi/pull/11205, as mentioned in https://github.com/highfidelity/hifi/pull/11205#issuecomment-328642626

- Fixes the issue where grabbers would re-appear when rotating the selected object rather than staying hidden.
- Fixes a bug with translateXZTool where its startingElevation and startingDistance properties were only set when local debug var was set. This appears to have been responsible for being able to move the object farther than was visible. Re-tested with fix and wasn't able to get that behavior any longer.
- This fixes an issue with multi-selection object rotation for pitch & roll that was previously exclusive to yaw rotation tool and also dedupes rotation tool code between the 3 tools where able.
- Fixes situations where attempting to click on a rotate, grab, or scale handle that was in front of others might unexpectedly activate the obscured handle.
- Fixes rotation skip near 0 and 180 degrees "poles"
- Fixes issue where the user view was obstructed by grabber handles when translating the selected object(s).
- Cleaned up onMouseEvents and moved all functionality for tools within the tools themselves as opposed to being split between the tool & which ever mouse event.

For more details, please refer to contained commit messages.

**Testing:**

1. Enter a domain where you have editing/rez permissions.
2. Create and/or select objects and perform the following operations with both Single & Multi-selection:
    - Rotation
    - Translation
    - Scaling
    - Cloning
3. The behavior should be inline with the expected results below.

**Expected Behavior:**

**Rotation:**
- While rotating longer the outer circle overlay, the user should see fine grain degree changes even around the 0 & 180 degree points.
- Multi-selection rotation along the Pitch & Roll, like Yaw, should now have proper solar system type rotation. (Note:  The "sun" position is the center of bounding box which encapsulates _all_ items within the selection).
- Upon clicking a rotation handle, all rotation, translation, scaling, & grabber handles should turn and remain invisible until the user releases the mouse press.

**Translation:**
- When moving the object along the z-axis:
    - The object doesn't warp behind the user/camera when moving the selection towards themselves or into infinity when moving the selection forward into the screen.
    - All rotation, translation, scaling, & grabber handles should turn invisible on click and remain invisible until the user releases the mouse press.
- When moving the object along the x-axis:
    - It's no longer possible to warp the selection into infinity.
    - All rotation, translation, scaling, & grabber handles should turn invisible on click and remain invisible until the user releases the mouse press.
- When moving the object along the y-axis:
    - It's no longer possible to warp the selection into infinity.
    - The vertical translation handle should remain visible when clicked and stay visible during the translation.  When the user releases the mouse press, the vertical translation handle should remain visible.
    - Excluding the vertical translation handle, all other rotation, translation, scaling, & grabber handles should turn invisible on clicking the vertical translation handle.  All the invisible handles should remain invisible until the user releases the mouse press.

**Multi-Selection:**
- All translation and rotation behavior noted above should also apply to multi-selection.

**General:**
- When clicking on a pair of overlapping handles, the handle which is directly under the cursor/ray should activate as opposed to the last tool checked.
